src/services/download/http.rs

```diff
@@ -8,35 +8,50 @@ use std::time::Duration;
 
 use actix_web::http::header;
 use actix_web::{client, HttpMessage};
+use anyhow::Result;
 use futures::compat::Stream01CompatExt;
 use futures::prelude::*;
 use url::Url;
 
-use super::{DownloadError, DownloadStatus, USER_AGENT};
-use crate::sources::{FileType, HttpSourceConfig, SourceFileId, SourceLocation};
-use crate::types::ObjectId;
+use super::locations::join_url_encoded;
+use super::{DownloadError, DownloadStatus, ObjectFileSource, SourceLocation, USER_AGENT};
+use crate::sources::{FileType, HttpSourceConfig};
+use crate::types::{ObjectFileSourceURI, ObjectId};
 use crate::utils::futures as future_utils;
 use crate::utils::http;
 
 /// The maximum number of redirects permitted by a remote symbol server.
 const MAX_HTTP_REDIRECTS: usize = 10;
 
-/// Joins the relative path to the given URL.
-///
-/// As opposed to [`Url::join`], this only supports relative paths. Each segment of the path is
-/// percent-encoded. Empty segments are skipped, for example, `foo//bar` is collapsed to `foo/bar`.
-///
-/// The base URL is treated as directory. If it does not end with a slash, then a slash is
-/// automatically appended.
-///
-/// Returns `Err(())` if the URL is cannot-be-a-base.
-fn join_url_encoded(base: &Url, path: &SourceLocation) -> Result<Url, ()> {
-    let mut joined = base.clone();
-    joined
-        .path_segments_mut()?
-        .pop_if_empty()
-        .extend(path.segments());
-    Ok(joined)
+/// The HTTP-specific [`ObjectFileSource`].
+#[derive(Debug, Clone)]
+pub struct HttpObjectFileSource {
+    pub source: Arc<HttpSourceConfig>,
+    pub location: SourceLocation,
+}
+
+impl From<HttpObjectFileSource> for ObjectFileSource {
+    fn from(source: HttpObjectFileSource) -> Self {
+        Self::Http(source)
+    }
+}
+
+impl HttpObjectFileSource {
+    pub fn new(source: Arc<HttpSourceConfig>, location: SourceLocation) -> Self {
+        Self { source, location }
+    }
+
+    pub fn uri(&self) -> ObjectFileSourceURI {
+        match self.url() {
+            Ok(url) => url.as_ref().into(),
+            Err(_) => "".into(),
+        }
+    }
+
+    /// Returns the URL from which to download this object file.
+    pub fn url(&self) -> Result<Url> {
```

Comment:
I would prefer to make the signature `Result<Url, anyhow::Error>`, since it's easier to see at a glance what this returns. That said, since this is a `join_url` operation, can we make this return a URL-specific error? I'd rather have the caller coerce this.

Reply:
I thought our current approach was to not use custom errors unless we need to match on them?

The `Result<Url>` part is more just familiarity imho, once you're used to `anyhow::Result` it's a pretty obvious shortcut.

Reply:
So if you write `anyhow::Result` rather than importing, I'm also fine with it 👍  

Point taken, let's go with anyhow for now. I'll propose something for this in our style guide. What I've often seen the community do, is:

- If there is a function that performs a single task, then throw the specific error for this task. For example, I/O functions should throw `std::io::Error`. Parse functions should throw a custom parse error.
- For propagating, libraries should create their own error type, either with a plain `ErrorKind` enum or with flag getters.
- For propagating, applications may use a boxed error such as anyhow as long as no error *handling* is needed.